scipy/special/xsf/tools.h

```diff
@@ -261,10 +258,169 @@ namespace detail {
+     * which the signs of f(a) and f(b) differ.
+     *
+     * This is a private function intended specifically for the situation where
+     * the goal is to invert a CDF function F for a parametrized family of distributions with respect to one
+     * parameter, when the other parameters are known, and where F is monotonic with respect to the unknown parameter.
+     *
+     * Note that this takes a pointer to a function taking a tuple of args along with a scalar
+     * double argument. A tuple of args for specializing func is also passed as the final argument.
+     * It would be much cleaner to use std::function and capturing lambda's to specialize the
+     * function we are finding a root for, but this was found unworkable in CuPy using NVRTC.
+     * This should be revisited in the future in order to allow simplifying this code. */
+    template <typename... Args>
+    XSF_HOST_DEVICE inline std::tuple<double, double, double, double, int> bracket_root_for_cdf_inversion(
+	double (*func)(double, std::tuple<Args...>), double x0, double xmin, double xmax, double step0_left, double step0_right,
+        double factor_left, double factor_right, bool increasing, std::uint64_t maxiter, std::tuple<Args...> args
+    ) {
+        double y0 = func(x0, args);
+
+	if (y0 == 0) {
+	    // Initial guess is correct.
+	    std::tuple<double, double, double, double, int> result(x0, x0, y0, y0, 0);
+	    return result;
+	}
+
+        double y0_sgn = std::signbit(y0);
+
+        bool search_left;
+        /* The frontier is the new leading endpoint of the expanding bracket. The
+         * interior endpoint trails behind the frontier. In each step, the old frontier
+         * endpoint becomes the new interior endpoint. */
+        double interior, frontier, y_interior, y_frontier, y_interior_sgn, y_frontier_sgn, boundary, factor;
+        if ((increasing && y0 < 0) || (!increasing && y0 > 0)) {
+            /* If func is increasing  and func(x_right) < 0 or if func is decreasing and
+             *  f(y_right) > 0, we should expand the bracket to the right. */
+            interior = x0, y_interior = y0;
+	    frontier = x0 + step0_right;
+	    y_frontier = func(frontier, args);
+            y_interior_sgn = y0_sgn;
+            y_frontier_sgn = std::signbit(y_frontier);
+            search_left = false;
+            boundary = xmax;
+	    factor = factor_right;
+        } else {
+	    /* Otherwise we move and expand the bracket to the left. */
+	    interior = x0, y_interior = y0;
+	    frontier = x0 + step0_left;
+	    y_frontier = func(frontier, args);
+            y_interior_sgn = y0_sgn;
+            y_frontier_sgn = std::signbit(y_frontier);
+            search_left = true;
+            boundary = xmin;
+	    factor = factor_left;
+        }
+
+        bool reached_boundary = false;
+        for (std::uint64_t i = 0; i < maxiter; i++) {
+	    if (y_frontier_sgn != y_interior_sgn || (y_frontier == 0.0)) {
```

Comment:
Would you like to branch for the case `y_frontier == 0.0` and return `(frontier, frontier)`? That would make it consistent with the treatment of `y0` and possibly save some future work for the caller.

Reply:
I think it's fine like this. The caller checks both endpoints anyway and it looks cleaner like this imo. The consistent principle I'm trying to go with is "do the simplest thing that returns a bracket that the caller will recognize as having a zero on an endpoint."